src/lin_diffEq_HHL.jl

```diff
@@ -32,52 +74,48 @@ function prepare_init_state(tspan::Tuple,x::Vector,h::Float64,g::Function)
   init_state
 end
 
-function Array_QuEuler(tspan::Tuple,h::Float64,g::Function)
-  N_t = round(2*(tspan[2] - tspan[1])/h + 3)
+function Array_QuDiff(tspan::Tuple,h::Float64,g::Function,alg::alg_lin_diffeq)
+  N_t = Int(round(2*(tspan[2] - tspan[1])/h + 3))
+  sz = size(g(1))
   I_mat = Matrix{Float64}(I, size(g(1)));
   A_ = I_mat;
-  zero_mat = zero(I_mat);
-  tmp_A = Array{Float64,2}(undef,size(g(1)))
+  # Generates First two rows
+  A_ = [A_ zeros(sz[1],sz[1]*(N_t));-1*(I_mat + h*g(tspan[1])) I_mat zeros(sz[1],sz[1]*(N_t-1))]
+  #Generates additional rows based on k - step
+  if alg.step > 1
+    A_ = [A_; Aval(QuAB2,2*h + tspan[1],h,g) zeros(sz[1],sz[1]*(N_t-2))]
+  end
 
-  for j = 2: N_t +1
-    A_ = [A_ zero_mat]
+  for i = 3:alg.step-1
+    tA_ = Aval(QuAB2,(i-1)*h + tspan[1],h,g)
+    tA_ = [zeros(sz[1],sz[1]*(i-2)) tA_]
+    A_ = [A_; tA_ zeros(sz[1],sz[1]*(N_t-i))]
   end
 
-  for i = 2 : N_t+1
-    tmp_A =  g(i*h + tspan[1])
-    tmp_A = -1*(I_mat + h*tmp_A)
-    tA_ = copy(zero_mat)
-    if i<3
-      tA_ = [tmp_A I_mat]
-    else
-      for j = 1:i-3
-        tA_ = [tA_ zero_mat]
-      end
-      if i < (N_t + 1)/2 + 1
-        tA_ = [tA_ tmp_A I_mat]
-      else
-        tA_ = [tA_ -1*I_mat I_mat]
-      end
-    end
-    if i< N_t+1
-      for j = i+1: N_t+1
-        tA_ = [tA_ zero_mat]
-      end
-    end
-    A_ = [A_;tA_]
+  if alg.step > 2
+    A_ = [A_;Aval(alg,(alg.step)*h + tspan[1],h,g) zeros(sz[1],sz[1]*(N_t-alg.step)) ]
+  end
 
+  for i = alg.step+1: (N_t + 1)/2 -2
+    tA_ = Aval(alg,(i-1)*h + tspan[1],h,g)
+    A_= [A_;zeros(sz[1],Int(sz[1]*(i-alg.step))) tA_ zeros(sz[1],Int(sz[1]*(N_t-i)))]
+   end
+  #Generates half mirroring matrix
+  for i = (N_t + 1)/2  : N_t
+    A_ = [A_; zeros(sz[1],Int(sz[1]*(i-2))) -1*I_mat I_mat zeros(sz[1],Int(sz[1]*(N_t + 1 -i)))]
   end
-    A_ = [zero(A_) A_;A_' zero(A_)]
-    A_
+  A_ = [A_; zeros(sz[1],sz[1]*(N_t-1)) -1*I_mat I_mat];
+
+  A_ = [zero(A_) A_;A_' zero(A_)]
+  A_
 end
 
-function solve_QuEuler(A::Function, b::Function, x::Vector,tspan::Tuple, h::Float64)
+function solve_QuDiff(alg::alg_lin_diffeq ,A::Function, b::Function, x::Vector,tspan::Tuple, h::Float64, n_reg::Int)
```

Comment:
I guess `solve` is the proper interface name for DifferentialEquations? Use a type to dispatch this quantum solve algorithm would make this looks more elegant?

FYI: http://docs.juliadiffeq.org/latest/tutorials/ode_example.html#Choosing-a-Solver-Algorithm-1

Reply:
I have made some changes. Please let me know they are okay.

Reply:
About this. Should I try to make use of the architecture present in DifferentialEquations? 
1. Define the problem as an `ODEProblem`.
2. Have `solve` use the HHL algorithm.